runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,403 @@
+import org.apache.beam.sdk.util.WindowingInternals;
+import org.apache.beam.sdk.util.state.StateInternals;
+import org.apache.beam.sdk.util.state.StateNamespaces;
+import org.apache.beam.sdk.util.state.StateTag;
+import org.apache.beam.sdk.util.state.StateTags;
+import org.apache.beam.sdk.util.state.ValueState;
+import org.apache.beam.sdk.util.state.WatermarkHoldState;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.TupleTag;
+import org.joda.time.Instant;
+
+/**
+ * A utility transform that executes a <a
+ * href="https://s.apache.org/splittable-do-fn">splittable</a> {@link DoFn} by expanding it into a
+ * network of simpler transforms: pair with restriction, split restrictions, assign unique key and
+ * group by it (in order to get access to per-key state and timers), then process.
+ *
+ * <p>This transform is intended as a helper for internal use by runners when implementing {@code
+ * ParDo.of(splittable DoFn)}, but not for direct use by pipeline writers.
+ */
+public class SplittableParDo<
+        InputT, OutputT, RestrictionT, TrackerT extends RestrictionTracker<RestrictionT>>
+    extends PTransform<PCollection<InputT>, PCollection<OutputT>> {
+  private final GroupByKeyIntoKeyedWorkItems<String, KV<InputT, RestrictionT>> groupByKeyTransform;
+  private final DoFn<InputT, OutputT> fn;
+
+  /**
+   * Creates the transform for the given original {@link ParDo} and {@link DoFn}, using a
+   * runner-specific implementation of {@link GroupByKeyIntoKeyedWorkItems}.
+   *
+   * @param name Name for the transform
+   * @param fn The splittable {@link DoFn} inside the original {@link ParDo} transform.
+   * @param groupByKeyTransform A runner-specific {@link GroupByKeyIntoKeyedWorkItems}.
+   */
+  public SplittableParDo(
+      String name,
+      DoFn<InputT, OutputT> fn,
+      GroupByKeyIntoKeyedWorkItems<String, KV<InputT, RestrictionT>> groupByKeyTransform) {
+    super(name);
+    this.groupByKeyTransform = groupByKeyTransform;
+    this.fn = fn;
+  }
+
+  @Override
+  public PCollection<OutputT> apply(PCollection<InputT> input) {
+    PCollection.IsBounded isFnBounded =
+        DoFnSignatures.INSTANCE.getOrParseSignature(fn.getClass()).isBounded();
+    Coder<RestrictionT> restrictionCoder =
+        DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn).invokeGetRestrictionCoder();
+    KvCoder<InputT, RestrictionT> splitCoder = KvCoder.of(input.getCoder(), restrictionCoder);
+
+    return input
+        .apply(
+            "Pair with initial restriction",
+            ParDo.of(new PairWithRestrictionFn<InputT, OutputT, RestrictionT>(fn)))
+        .setCoder(splitCoder)
+        .apply("Split restriction", ParDo.of(new SplitRestrictionFn<InputT, RestrictionT>(fn)))
+        .apply(
+            "Assign unique key", ParDo.of(new AssignRandomUniqueKeyFn<KV<InputT, RestrictionT>>()))
```

Comment:
I don't think we should require restrictions to have an ID. They're arbitrary simple mathematical objects, we can generate an ID ourselves without requiring the user to.